pkg/storage/wal/manager.go

```diff
@@ -0,0 +1,220 @@
+	cfg Config
+
+	// available is a list of segments that are accepting writes. When a
+	// segment is full, or the close timer has expired, it is moved to the
+	// closed list to be flushed.
+	available *list.List
+
+	// closed is a list of segments that are full and waiting to be flushed.
+	// Once flushed, the segment is moved back to the available list to
+	// accept writes again.
+	closed   *list.List
+	shutdown chan struct{}
+	mu       sync.Mutex
+}
+
+// item is similar to ClosedWriter, but it is an internal struct used in the
+// available and closed lists. It contains a single-use result that is returned
+// to callers of Append and a re-usable writer that is reset after each flush.
+type item struct {
+	r *AppendResult
+	w *SegmentWriter
+}
+
+// ClosedWriter contains a result and a writer that has either exceeded the
+// maximum age or the maximum size for a segment. ClosedWriters are to be
+// flushed and then returned so the writer can be re-used.
+type ClosedWriter struct {
+	Result *AppendResult
+	Writer *SegmentWriter
+}
+
+func NewManager(cfg Config) (*Manager, error) {
+	m := Manager{
+		cfg:       cfg,
+		available: list.New(),
+		closed:    list.New(),
+		shutdown:  make(chan struct{}),
+	}
+	for i := int64(0); i < cfg.MaxSegments; i++ {
+		w, err := NewWalSegmentWriter()
+		if err != nil {
+			return nil, err
+		}
+		m.available.PushBack(&item{
+			r: &AppendResult{done: make(chan struct{})},
+			w: w,
+		})
+	}
+	return &m, nil
+}
+
+func (m *Manager) Append(r AppendRequest) (*AppendResult, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	// When the WAL becomes full it stops accepting writes.
+	if m.available.Len() == 0 {
+		return nil, ErrFull
+	}
+	el := m.available.Front()
+	it := el.Value.(*item)
+	it.w.Append(r.TenantID, r.LabelsStr, r.Labels, r.Entries)
+	// If the item exceeded the maximum age or the maximum size, move it to
+	// closed to be flushed.
+	if it.w.InputSize() >= m.cfg.MaxSegmentSize {
+		m.closed.PushBack(it)
+		m.available.Remove(el)
+	}
+	return it.r, nil
+}
+
+func (m *Manager) Get() (*ClosedWriter, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.closed.Len() == 0 {
+		return nil, nil
+	}
+	el := m.closed.Front()
+	it := el.Value.(*item)
+	m.closed.Remove(el)
+	return &ClosedWriter{Result: it.r, Writer: it.w}, nil
+}
+
+func (m *Manager) Put(it *ClosedWriter) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	it.Writer.Reset()
+	m.available.PushBack(&item{
+		r: &AppendResult{done: make(chan struct{})},
+		w: it.Writer,
+	})
+	return nil
+}
+
+// Run the Manager. This runs a number of periodic tasks.
+func (m *Manager) Run() error {
+	t := time.NewTicker(m.cfg.MaxAge)
+	for {
+		select {
+		case <-t.C:
+			m.doTick()
+		case <-m.shutdown:
+			return nil
+		}
+	}
+}
+
+// Stop the Manager.
+func (m *Manager) Stop() error {
+	close(m.shutdown)
+	return nil
+}
+
+func (m *Manager) doTick() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	// Check if the current segment has exceeded its maximum age, and if so,
```

Comment:
This isn't actually being checked here, we just check if the first segment has any data and assume it'll be old enough to flush. It might be a very small segment if we run a tick just after hitting a size threshold in Append

Reply:
Good call !